controllers/helpers.go

```diff
@@ -113,3 +121,244 @@ func GetObjectsToRequestsByNamespaceAndClusterName(ctx context.Context, c client
 	}
 	return results
 }
+
+// Returns true if a and b point to the same object
+func referSameObject(a, b metav1.OwnerReference) bool {
+	aGV, err := schema.ParseGroupVersion(a.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	bGV, err := schema.ParseGroupVersion(b.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	return aGV.Group == bGV.Group && a.Kind == b.Kind && a.Name == b.Name
+}
+
+// GetCloudProviderSecret returns the required azure json secret for this machine template
+func GetCloudProviderSecret(d azure.ClusterDescriber, namespace, name string, identityType infrav1.VMIdentity, userIdentityID string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      fmt.Sprintf("%s-azure-json", name),
+		},
+	}
+
+	var controlPlaneConfig, nodeConfig []byte
+	var err error
+
+	switch identityType {
+	case infrav1.VMIdentitySystemAssigned:
+		controlPlaneConfig, err = systemMSIControlPlaneCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig, err = systemMSINodeCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityUserAssigned:
+		if len(userIdentityID) < 1 {
+			return nil, errors.New("expected a non-empty userIdentityID")
+		}
+		controlPlaneConfig, err = userMSIControlPlaneCloudProviderConfig(d, userIdentityID)
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig, err = userMSINodeCloudProviderConfig(d, userIdentityID)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityNone:
+	default:
+		controlPlaneConfig, err = servicePrincipalControlPlaneCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig, err = servicePrincipalNodeCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	secret.Data = map[string][]byte{
+		"controlplane": controlPlaneConfig,
+		"node":         nodeConfig,
```

Comment:
what are your thoughts on potentially needing to split up the azure.json for node into one for each machine deployment / machine pool in the future? Thinking of use cases where one of the configurations would change per node pool such as having multiple subnets

Reply:
that's what this does already, it just always generates both controlplane and node versions for even when the MD can only be one or the other. we differentiate by referencing the desired key in the MD kubeadm config

Reply:
I guess it's pretty easy to use the control plane label to tell them apart and only use one key

Reply:
we don't actually plumb the per-machinepool subnet through right now because we get it from the cluster spec. But if we switched that to e.g. the machine spec, we could easily adjust the interfaces/params here to make it work.

Reply:
sorry for stream of consciousness -- I think I remembered the reason I didn't differentiate. 

For machines, you can tell if they're control plane, machine set, or machine deployment based. If it's not any of those (i.e., a user created a standalone machine for some reason), you want the catch that and create a single-machine azure json.

For machine templates, the cluster is the owner. We would have to go back to the cluster, get the control plane ref, fetch that object, and check if the control plane uses this machine template or not. BUT there's not really a strict guarantee that the infra ref is KCP, so we can't use our knowledge of KCP schema to look up the azure machine template spec. 

The other alternative that would allow is creating azure-json per control plane machine instead of per-machine template for the control plane. That's strictly worse as we end up generating N times as many secrets as required (where N == control plane replicas), vs duplicating the data one time inside the secret data map.